KAFKA-2812: improve consumer integration tests

hachikuji opened the pull request.

hachikuji committed:
KAFKA-2812: improve consumer integration tests

hachikuji commented:
@granders Ping for initial review.

hachikuji committed:
minor fix to allow tests to work with more consumers

hachikuji committed:
add consumer/broker bounce tests

guozhangwang commented:
LGTM, leave to @granders to take a look.

hachikuji commented:
@guozhangwang Thanks for taking a look. I've also got some bounce tests that I'll probably add later today. They helped to find the deadlock error.

granders commented on tests/kafkatest/services/verifiable_consumer.py:
```diff
@@ -92,55 +174,40 @@ def _worker(self, idx, node):
-            partition = topic_partition["partition"]
-            assignment.append(TopicPartition(topic, partition))
-        self.assignment[node] = assignment
+            with self.lock:
```
Why are you able to remove this `None` check now?

hachikuji replied:
It should still be there. Thanks for catching!

granders commented on tests/kafkatest/services/verifiable_consumer.py:
```diff
@@ -92,55 +174,40 @@ def _worker(self, idx, node):
+            tp = TopicPartition(consumed_partition["topic"], consumed_partition["partition"])
+            if tp in self.global_committed:
+                # verify that the position never gets behind the current commit.
+                # it would be nice to also verify that the position increases monotonically
```
Even if you can't guarantee this in case of hard client failures, perhaps you could still have a check and log it (warning level?)

hachikuji replied:
Yeah, good point.

granders commented on tests/kafkatest/services/verifiable_consumer.py:
```diff
@@ -174,6 +241,10 @@ def try_parse_json(self, string):
             return None
 
+    def stop_all(self):
+        for node in self.event_handlers:
```
I think this is equivalent to `for node in self.nodes`, which might be clearer

granders commented on tests/kafkatest/services/verifiable_consumer.py:
```diff
@@ -203,20 +273,47 @@ def clean_node(self, node):
     def current_assignment(self):
         with self.lock:
-            return self.assignment
+            return dict((handler.node, handler.current_assignment()) for handler in self.event_handlers.itervalues())
```
Super minor, but you can also express this as
`return {handler.node: handler.current_assignment() for handler in self.event_handlers.values()}`

granders commented on tests/kafkatest/services/verifiable_consumer.py:
```diff
@@ -20,12 +20,94 @@
+        for record_batch in event["partitions"]:
+            tp = TopicPartition(topic=record_batch["topic"],
+                                partition=record_batch["partition"])
+            minOffset = record_batch["minOffset"]
```
Also minor
`minOffset` -> `min_offset`
`maxOffset` -> `max_offset`

granders commented:
@hachikuji A few very minor comments. Otherwise looks good

hachikuji committed:
fix incorrect comment

hachikuji commented:
@granders, @guozhangwang: I added some bounce tests and addressed Geoff's comments. Unless there are any concerns, this is probably ready to merge whenever there is a good opportunity.
